lib/chef/provisioning/docker_driver/driver.rb

```diff
@@ -70,66 +70,116 @@ def self.connection_url(driver_url)
+        container_id = machine_spec.reference['container_id']
+        image_id ||= machine_spec.reference['image_id']
+        docker_options ||= machine_spec.reference['docker_options']
+      end
 
-      container_name = machine_spec.name
+      container_name ||= machine_spec.name
       machine_spec.reference = {
-          'driver_url' => driver_url,
-          'driver_version' => Chef::Provisioning::DockerDriver::VERSION,
-          'allocated_at' => Time.now.utc.to_s,
-          'host_node' => action_handler.host_node,
-          'container_name' => container_name,
-          'image_id' => machine_options[:image_id],
-          'docker_options' => machine_options[:docker_options]
+        'driver_url' => driver_url,
+        'driver_version' => Chef::Provisioning::DockerDriver::VERSION,
+        'allocated_at' => Time.now.utc.to_s,
+        'host_node' => action_handler.host_node,
+        'container_name' => container_name,
+        'image_id' => image_id,
+        'docker_options' => docker_options,
+        'container_id' => container_id
       }
+      build_container(machine_spec, docker_options)
     end
 
     def ready_machine(action_handler, machine_spec, machine_options)
-      base_image_name = build_container(machine_spec, machine_options)
-      start_machine(action_handler, machine_spec, machine_options, base_image_name)
-      machine_for(machine_spec, machine_options, base_image_name)
+      start_machine(action_handler, machine_spec, machine_options)
+      machine_for(machine_spec, machine_options)
     end
 
-    def build_container(machine_spec, machine_options)
-      docker_options = machine_options[:docker_options]
+    def build_container(machine_spec, docker_options)
+      container = container_for(machine_spec)
+      return container unless container.nil?
+
+      image = find_image(machine_spec) ||
+        build_image(machine_spec, docker_options)
+
+      args = [
+        'docker',
+        'run',
+        '--name',
+        machine_spec.reference['container_name'],
+        '--detach'
+      ]
+
+      if docker_options[:keep_stdin_open]
+        args << '-i'
+      end
+
+      if docker_options[:env]      
+        docker_options[:env].each do |key, value|
+          args << '-e'
+          args << "#{key}=#{value}"
+        end
+      end
+
+      if docker_options[:ports]
+        docker_options[:ports].each do |portnum|
+          args << '-p'
+          args << "#{portnum}"
+        end
+      end
 
-      base_image = docker_options[:base_image]
-      if !base_image
-        Chef::Log.debug("No base images specified in docker options.")
-        base_image = base_image_for(machine_spec)
+      if docker_options[:volumes]
+        docker_options[:volumes].each do |volume|
+          args << '-v'
+          args << "#{volume}"
+        end
       end
+
+      args << image.id
+      args += Shellwords.split("/bin/sh -c 'while true;do sleep 1; done'")
```

Comment:
what are your concerns? changing this to running interactively fundamentally changes how this all flows. I don't think `cmd.run_command` would simply return and leave you with a running container.